modules/fvad/fvad.c

```diff
@@ -0,0 +1,317 @@
+	}
+
+	list_unlink(&st->af.le);
+}
+
+
+static void dec_destructor(void *arg)
+{
+	struct vad_dec *st = arg;
+
+	if (st->fvad) {
+		fvad_free(st->fvad);
+	}
+
+	list_unlink(&st->af.le);
+}
+
+
+static void print_vad(int pos, int color, bool tx, bool active)
+{
+	/* move cursor to a fixed position */
+	re_fprintf(stderr, "\x1b[%dG", pos);
+
+	/* print vad in Nice colors */
+	re_fprintf(stderr, " \x1b[%dm[%s]\x1b[;m\r",
+		   color,  active ? (tx ? "tx" : "rx") : "  ");
+}
+
+
+static int encode_update(struct aufilt_enc_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_enc *st;
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	st->au = au;
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	return 0;
+}
+
+
+static int decode_update(struct aufilt_dec_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_dec *st;
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), dec_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	st->au = au;
+
+	*stp = (struct aufilt_dec_st *)st;
+
+	return 0;
+}
+
+
+static bool auframe_vad(Fvad *fvad, struct auframe *af)
+{
+	static int chunk_times_ms[] = { 30, 20, 10 };
+
+	int16_t *buf = NULL;
+	int16_t *allocated = NULL;
+	bool detected = false;
+
+	/* convert to 16 bit linear */
+	if (af->fmt == AUFMT_S16LE) {
+		buf = af->sampv;
+	}
+	else {
+		allocated = (int16_t*)mem_alloc(sizeof(int16_t) * af->sampc,
+			NULL);
+		buf = allocated;
+		auconv_to_s16(buf, af->fmt, af->sampv, af->sampc);
+	}
+
+	size_t ms = af->sampc * 1000 / af->srate;
+	size_t pos = 0;
+
+	/* process all chunk_sizes that fvad accepts */
+	for (size_t chunk_time_index = 0;
+		chunk_time_index < RE_ARRAY_SIZE(chunk_times_ms);
+		++chunk_time_index) {
+
+		const size_t chunk_time = chunk_times_ms[chunk_time_index];
+		const size_t sampc = af->srate / 1000 * chunk_time;
+
+		for (size_t i = 0; i < ms / chunk_time; ++i) {
+			int err = fvad_process(fvad, buf + pos, sampc);
+			pos += sampc;
+			ms -= chunk_time;
+			if (err > 0) {
+				detected = true;
+				goto out;
+			}
+			else if (err < 0) {
+				warning("fvad_process(%d) failed", sampc);
+				goto out;
+			}
+		}
+	}
+
+	if (pos != af->sampc) {
+		warning("fvad_process: samples left over: %d",
+			af->sampc - pos);
+	}
+
```

Comment:
I don't like this. When the ptime is smaller than 10ms, the VAD library cannot process frames.

I see two possibilities:
- return EINVAL when the ptime is smaller than 10ms (this would require writing something like `int audio_get_ptime(struct audio *au, bool tx)`
- store leftover data and process it with the next frame

Thoughts?

Reply:
You could try with `mbuf` to collect 20ms frames.